### ISSUE TITLE
Compatibility with Gitlab test reports

### ISSUE THREAD

**Issue by JackDunnNZ** [JuliaTesting/TestReports.jl#43]
Thanks for the package! I was playing around trying to integrate it with the [unit test reporting feature in gitlab](https://docs.gitlab.com/ee/ci/unit_test_reports.html), but I discovered that there are some quirks as to how gitlab is processing the files that means it doesn't work out of the box.

Basically, it doesn't look at any of the information in `testsuite` and just uses `testcase`. It uses the combination of the `classname` and `name` attributes on a `testcase` to uniquely identify the testcases for reporting and display.

For instance, the following output

```xml
<testsuites>
    <testsuite>
        <testcase classname='Calculator' name='sumTest1' time='0.01'></testcase>
        <testcase classname='Calculator' name='sumTest1' time='0.02'></testcase>
    </testsuite>
    <testsuite>
        <testcase classname='Statemachine' name='sumTest1' time='100'></testcase>
        <testcase classname='Statemachine' name='unhappy path' time='200'></testcase>
    </testsuite>
</testsuites>
```

is displayed in the UI as

![image](https://user-images.githubusercontent.com/4717044/94368881-a0fd4c00-00b4-11eb-9e7c-41084c40e527.png)

You can see that the first case in the first suite is dropped as it appears to be overwritten by the second one, due to the same `classname`/`name`. The Statemachine/sumTest1 does not conflict, indicating the classname and name are both being used.

I identified that this is the relevant part of the parsing code in Gitlab to see which fields they are using from `testcase`:

https://gitlab.com/gitlab-org/gitlab-foss/-/blob/8634097176560e9febf4ef6b04100427787e21e0/lib/gitlab/ci/parsers/test/junit.rb#L49-75

Right now, the output from this package looks like

```xml
<testsuites name="" id="_id_" tests="3" failures="0" errors="0">
   <testsuite name="testset1/testset2/testset3" id="_id_" tests="2" failures="0" errors="0">
      <testcase name="pass (info lost)" id="_testcase_id_" />
      <testcase name="pass (info lost)" id="_testcase_id_" />
   </testsuite>
   <testsuite name="testset1/testset2/testset4" id="_id_" tests="1" failures="0" errors="0">
      <testcase name="pass (info lost)" id="_testcase_id_" />
   </testsuite>
</testsuites>
```

Since there are no classnames, and all the names are the same, Gitlab merges them all into just a single testcase.

I wanted to raise this to see if there were any thoughts about how feasible it would be to set unique classname/name combinations for each testcase in the output. 

On a surface level, it feels like the testcase `classname` could just be the name of the testsuite, and the `name` could include an incremented counter?

That said, I don't know anything about the internals of how this package works, and definitely don't want to ask for drastic changes to accomodate this! If it doesn't fit here, it seems like I could easily achieve what is needed with a simple post-processing of the file.

**Comment by mmiller-max** [comment]
Thanks for raising this issue!

Looking at the [JUnit schema](https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd), it seems that `classname` is required, and therefore it's something we should add to `TestReports` as other CI tools may also be relying on this.

Your suggestions for how to make them differ makes sense. Hopefully at some point we'll be able to improve the name of passing tests (it needs a change in `Test` itself) so just incrementing it for now makes sense, and the `classname` field doesn't make a huge amount of sense anyway as we're not testing classes in Julia in the same way as you would with other languages, so using the testsuite name seems sensible.

This would still cause testcases to overlap in Gitlab if you had testsets with the same name, but that doesn't make a lot of sense for someone to do and would likely be a mistake anyway.

This should be a reasonably straightforward change to `to_xml.jl`, I'll try and get to it soon or feel free to give it a go too!

**Comment by oxinabox** [comment]
cc @iamed2

**Comment by mmiller-max** [comment]
Closed by #44
